staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,497 @@
+		}
+	})
+
+	for _, fl := range aliasArgs.flags {
+		existingFlag := foundAliasCmd.Flag(fl.Name)
+		if existingFlag == nil {
+			return args, fmt.Errorf("invalid alias flag %s in alias %s", fl.Name, args[0])
+		}
+		if searchInArgs(existingFlag.Name, existingFlag.Shorthand, allShorthands, args) {
+			// Don't modify the value implicitly, if it is passed in args explicitly
+			continue
+		}
+		err = foundAliasCmd.Flags().Set(fl.Name, fl.Default)
+		if err != nil {
+			return args, fmt.Errorf("could not apply value %s to flag %s in alias %s err: %w", fl.Default, fl.Name, args[0], err)
+		}
+	}
+
+	if len(aliasArgs.prependArgs) > 0 {
+		// prependArgs defined in kuberc should be inserted after the alias name.
+		if commandIndex+1 >= len(args) {
+			// command is the last item, we simply append just like appendArgs
+			args = append(args, aliasArgs.prependArgs...)
+		} else {
+			args = append(args[:commandIndex+1], append(aliasArgs.prependArgs, args[commandIndex+1:]...)...)
+		}
+	}
+	if len(aliasArgs.appendArgs) > 0 {
+		// appendArgs defined in kuberc should be appended to actual args.
+		args = append(args, aliasArgs.appendArgs...)
+	}
+	// Cobra (command.go#L1078) appends only root command's args into the actual args and ignores the others.
+	// We are appending the additional args defined in kuberc in here and
+	// expect that it will be passed along to the actual command.
+	rootCmd.SetArgs(args[1:])
+	return args, nil
+}
+
+// DefaultGetPreferences returns KubeRCConfiguration.
+// If users sets kuberc file explicitly in --kuberc flag, it has the highest
+// priority. If not specified, it looks for in KUBERC environment variable.
+// If KUBERC is also not set, it falls back to default .kuberc file at the same location
+// where kubeconfig's defaults are residing in.
+// If KUBERC is set to "off", kuberc will be turned off and original behaviors in kubectl will be applied.
+func DefaultGetPreferences(kuberc string) (*config.Preference, error) {
+	if val := os.Getenv("KUBERC"); val == "off" {
+		if kuberc != "" {
+			return nil, fmt.Errorf("disabling kuberc via KUBERC=off and passing kuberc flag are mutually exclusive")
+		}
+		return nil, nil
+	}
+	kubeRCFile := RecommendedKubeRCFile
+	explicitly := false
+	if kuberc != "" {
+		kubeRCFile = kuberc
+		explicitly = true
+	}
+
+	if kubeRCFile == "" && os.Getenv("KUBERC") != "" {
+		kubeRCFile = os.Getenv("KUBERC")
+		explicitly = true
+	}
+
+	kubeRCBytes, err := os.ReadFile(kubeRCFile)
+	if err != nil {
+		if !explicitly {
+			// We don't log if the kuberc file does not exist. Because user simply does not
+			// specify neither default kuberc file nor explicitly pass it.
+			// We'll continue to default behavior without raising any error.
+			if !os.IsNotExist(err) {
+				klog.V(4).Infof("error reading kuberc file %q: %v", kubeRCFile, err)
+			}
+			return nil, nil
+		}
+		// Kuberc is specified by user via flag or env variable and it gets an error.
+		return nil, err
+	}
+
```

Comment:
This is getting closer, but we need to think more about how old kubectls and new kubectls will interoperate with a shared environment or shared ~/.kube/kuberc file, where the user wants to give v1-capable kubectl versions a v1 config and older kubectls a v1beta1 or v1alpha1 config

possibilities:
1. wait to enable KUBECTL_KUBERC by default until the config file reaches v1
2. (preferred) read a document stream from the kuberc file and take the first recognized object (tolerating undecodable objects as long as we get at least one workable object from an explicitly specified kuberc)

I think we should do the following:

1. loop through the kuberc file as an object _stream_
2. for each object in the stream:
   * try to decode strictly; if successful GVK decode, return the decoded object
   * if strict error, try to decode leniently; if successful, warn about the strict error and return the leniently decoded object
   * if lenient decode fails, remember the error
3. if we reach the end of the file and cannot decode any GVK successfully:
   * if the kuberc was explicitly requested, return accumulated decode errors
   * if the kuberc was implicitly requested, warn with accumulated decode errors

Reply:
kubeadm config looks sort of like this, though instead of v1 vs beta it's for different objects that are used for init vs join (this is the precedent I had in mind), the config format there is a yaml document stream of objects and only the relevant GVK are read.

kubeadm's is here I believe: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/componentconfigs/configset.go

kind has a DIY version but we'll want to use api-machinery proper ...

Reply:
Thank you @BenTheElder, example was very helpful. I think, I've applied successfully what @liggitt suggested above.